samples/image_viewer.cpp

```diff
@@ -0,0 +1,386 @@
+
+        if (filename.exists()) {
+            loadImage(app, engine, filename);
+        }
+    };
+
+    auto cleanup = [&app](Engine* engine, View*, Scene*) {
+        engine->destroy(app.scene.imageEntity);
+        engine->destroy(app.scene.imageVertexBuffer);
+        engine->destroy(app.scene.imageIndexBuffer);
+        engine->destroy(app.scene.imageMaterial);
+        engine->destroy(app.scene.imageTexture);
+        engine->destroy(app.colorGrading);
+
+        delete app.viewer;
+    };
+
+    auto gui = [&app](Engine* engine, View* view) {
+        app.viewer->updateUserInterface();
+
+        FilamentApp::get().setSidebarWidth(app.viewer->getSidebarWidth());
+    };
+
+    auto preRender = [&app](Engine* engine, View* view, Scene* scene, Renderer* renderer) {
+        auto& rcm = engine->getRenderableManager();
+
+        // This applies clear options, the skybox mask, and some camera settings.
+        Camera& camera = view->getCamera();
+        Skybox* skybox = scene->getSkybox();
+        applySettings(app.viewer->getSettings().viewer, &camera, skybox, renderer);
+
+        // Check if color grading has changed.
+        ColorGradingSettings& options = app.viewer->getSettings().view.colorGrading;
+        if (options.enabled) {
+            if (options != app.lastColorGradingOptions) {
+                ColorGrading *colorGrading = createColorGrading(options, engine);
+                engine->destroy(app.colorGrading);
+                app.colorGrading = colorGrading;
+                app.lastColorGradingOptions = options;
+            }
+            view->setColorGrading(app.colorGrading);
+        } else {
+            view->setColorGrading(nullptr);
+        }
+
+        if (app.scene.imageTexture) {
+            Texture *texture = app.scene.imageTexture;
+            float srcWidth = texture->getWidth();
+            float srcHeight = texture->getHeight();
+            float dstWidth = view->getViewport().width;
+            float dstHeight = view->getViewport().height;
+
+            float srcRatio = srcWidth / srcHeight;
+            float dstRatio = dstWidth / dstHeight;
+
+            bool xMajor = dstWidth / srcWidth > dstHeight / srcHeight;
+
+            float sx = 1.0f;
+            float sy = dstRatio / srcRatio;
+
+            float tx = 0.0f;
+            float ty = ((1.0f - sy) * 0.5f) / sy;
+
+            if (xMajor) {
+                sx = srcRatio / dstRatio;
+                sy = 1.0;
+                tx = ((1.0f - sx) * 0.5f) / sx;
+                ty = 0.0f;
+            }
+
+            mat3f transform(
+                 1.0f / sx,  0.0f,       0.0f,
+                 0.0f,       1.0f / sy,  0.0f,
+                -tx,        -ty,         1.0f
+            );
+
+            app.scene.imageMaterial->setDefaultParameter("transform", transform);
+        }
+    };
+
+    auto postRender = [&app](Engine* engine, View* view, Scene* scene, Renderer* renderer) {
+        // TODO
+    };
```

Comment:
You can remove this, FilamentApp supplies a default argument.

Reply:
Ah yeah I know, but I was originally planning on using it.